src/resources.rs

```diff
@@ -0,0 +1,125 @@
+// Copyright (C) 2019 Alibaba Cloud. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Structs to manage device resources.
+
+/// Enumeration for device resources.
+#[allow(missing_docs)]
+#[derive(Clone)]
+pub enum DeviceResourceEntry {
+    /// IO Port address range.
+    PioAddressRange { base: u16, size: u16 },
+    /// Memory Mapped IO address range.
+    MmioAddressRange { base: u64, size: u64 },
+    /// Legacy IRQ number.
+    LegacyIrq(u32),
+    /// PCI MSI IRQ numbers.
+    PciMsiIrq { base: u32, size: u32 },
+    /// PCI MSIx IRQ numbers.
+    PciMsixIrq { base: u32, size: u32 },
+    /// Generic MSI IRQ numbers.
+    GenericIrq { base: u32, size: u32 },
```

Comment:
Or even a MsiIrqType is needed at all?Depends on where we will use it.

Reply:
> Or even a MsiIrqType is needed at all?Depends on where we will use it.

A PCI device may support both PCI MSI and MSIx, and the guest kernel choose which one to use. So we need a way to distinguish different cases. If we assume only one type of MSI interrupt is supported by any device, we could remove the MSI type info.

Reply:
> Overall I believe all the source code under interrupt/kvm_irq should be part of VMM code instead of vm-device, in vm-device we don't need it at all. The only thing we do need is the trait definitions in interrupt/mod.rs

I agree the `kvm_irq` implementation is orthogonal to the traits definitions, but I also think that this code could be shared across VMM implementations through the vm-device crate, as a default KVM implementation of the interrupt traits. Kind of similar to what we do with the `mmap` implemenation of the memory model.

Reply:
We can (and probably should when considering device assignment) support multiple MSI types. From device backend side, it knows exactly which MSI type it exposed to guest so why it need the info here. What I'm seeing is this type is from what vm-device returned to VMM in ResourceRequest.